pyDeltaRCM/init_tools.py

```diff
@@ -33,31 +36,52 @@ def init_logger(self):
             # create the logging file handler
             st = timestr = time.strftime("%Y%m%d-%H%M%S")
             fh = logging.FileHandler("pyDeltaRCM_" + st + ".log")
-
             formatter = logging.Formatter(
                 '%(asctime)s - %(name)s - %(levelname)s - %(message)s')
             fh.setFormatter(formatter)
 
             # add handler to logger object
             self.logger.addHandler(fh)
 
-    def set_defaults(self):
+    def import_files(self):
+
+        # This dictionary serves as a container to hold values for both the
+        # user-specified file and the internal defaults.
+
+        input_file_vars = dict()
 
-        for k, v in list(self._var_default_map.items()):
-            setattr(self, self._var_name_map[k], v)
+        # Open and access both yaml files --> put in dictionaries
+        # only access the user input file if provided.
 
-    def create_dicts(self):
+        default_file = open(self.default_file, mode = 'r')
+        default_dict = yaml.load(default_file, Loader = yaml.FullLoader)
+        default_file.close()
+        if hasattr(self, 'input_file') and os.path.exists(self.input_file):
+            user_file = open(self.input_file, mode = 'r')
+            user_dict = yaml.load(user_file, Loader = yaml.FullLoader)
+            user_file.close()
+        else:
+            print('The specified input file could not be found. Using default values...')
+            user_dict = dict()
 
-        self._input_var_names = list(self._input_vars.keys())
+        # go through and populate input vars with user and default values,
+        # checking user values for correct type.
 
-        self._var_type_map = dict()
-        self._var_name_map = dict()
-        self._var_default_map = dict()
+        for oo in default_dict.keys():
+            the_type = eval(default_dict[oo]['type'])
+            if oo in user_dict and isinstance(user_dict[oo], the_type):
+                input_file_vars[oo] = user_dict[oo]
+            elif oo in user_dict and not isinstance(user_dict[oo], the_type):
+                print('Input for ' + oo + ' not of the right type. ' + \
+                oo + ' needs to be of type ' + str(the_type))
+                input_file_vars[oo] = default_dict[oo]['default']
+            else:
+                input_file_vars[oo] = default_dict[oo]['default']
 
-        for k in list(self._input_vars.keys()):
-            self._var_type_map[k] = self._input_vars[k]['type']
-            self._var_name_map[k] = self._input_vars[k]['name']
-            self._var_default_map[k] = self._input_vars[k]['default']
+        # now make each one an attribute of the model object.
+
+        for k, v in list(input_file_vars.items()):
```

Comment:
I see this code was already in the model. Is there a chance that an attribute already exists on `self`? We do not want to set it as an attr then.

For example, what if the user specifies something in the `.yaml` that isn't actually a supported option? and instead we set `self.entire_model=1` and break. 

I think your if-then tree above would be able to ignore these, but it is kind of hard to tell. I think some tests would be clutch here.

Reply:
Yeah the if-then tree should be able to do this, because it actually only imports attributes that are present in the `default_dict`, does that make sense? So if the user supplies `poop: 5000` in their input yaml, it won't match any of the designated model parameters in `default.yml`. If they want to bring `poop` into the model, they have to supply a default in the `default.yml` file.

Reply:
I verified this is correct with some of the tests in the PR I sent you.

I think it's probably fine that custom attributes can't be specified? Either way, we can address this with a future update if needed.